grants/speculative/visualize_rust_lifetime.md

```diff
@@ -0,0 +1,209 @@
+
+We divide the project into three milestones. We aim to finish the whole project in three months and achieve a milestone in each month.  
+
+
+* **Total Estimated Duration:** 3 months
+* **Full-time equivalent (FTE):**  3
+* **Total Costs:** financial information will be disclosed privately. 
+
+
+### Milestone 1 — Implement the bug detection component  
+* **Estimated Duration:** 1 month
+
+
+| Number | Deliverable | Specification |
+| ------------- | ------------- | ------------- |
+| 0a. | License | Apache 2.0 |
+| 0b. | Documentation | We will provide both inline documentation of the code and a basic tutorial that explains how to run the bug detection component as a standalone tool on terminal.|
+| 0c. | Testing Guide | We will include unit tests to ensure the functionality and robustness of our code. We will also include 10 toy programs containing different types of deadlocks to demonstrate the bug detection capability. We will also run this component on the latest version of Substrate, Polkadot, and ink!. We will manually inspect all reported results to count the number of bugs and the number of false positives. | 
+| 1. | Detecting Conflicting Locks  | We will implement a detector that can identify deadlocks due to locks in conflicting orders through analyzing the MIR of Rust programs.|  
+| 2. | Detecting Misuse of Mutex and Channel | We will implement a detector to identify deadlocks due to errors when using a mutex together with a channel. |  
+| 3. | Detecting Misuse of Mutex and Conditional Variable | We will implement a detector to identify deadlocks due to mistakes when using a mutex together with a conditional variable. | 
+| 4. | Docker | We will provide a dockerfile to demonstrate the full functionality of this component. |
+
+
+### Milestone 2 — Integrate the bug detection functionalities into RLS 
+* **Estimated Duration:** 1 month
+
+
+| Number | Deliverable | Specification |
+| ------------- | ------------- | ------------- |
+| 0a. | License | Apache 2.0 |
+| 0b. | Documentation | We will provide both inline documentation of the code and a basic tutorial that explains how to install and use the changed RLS.  |
+| 0c. | Testing Guide | We will also use the 10 toy programs designed in the last milestone to test whether the bug detection capability is successfully integrated into RLS. | 
+| 1. | Extend Language Server Protocol (LSP) | We will extend LSP to contain the key information related to deadlocks (e.g., the start and the end of a critical section, blocking operations in a critical section).|  
+| 2. | Change RLS to emit MIR  | We will change the build module of RLS to emit MIR for our bug detection functionalities. |  
+| 3. | Conduct bug detection in RLS | We will change the analysis crate of RLS to execute the code for bug detection and change the server module to send the detection results out. |  
```

Comment:
Does rust-analyzer have an analysis crate and a server module, or otherwise what would be the equivalent here?

```suggestion
| 3. | Conduct bug detection in rust-analyzer | We will change the analysis crate of rust-analyzer to execute the code for bug detection and change the server module to send the detection results out. |  
```